### ISSUE TITLE
Google::Protobuf::DescriptorPool.generated_pool.build SEGV at 0x00000000000090

### ISSUE THREAD

**Issue by tarcieri** [google/protobuf#3649]
I've only seen it once while loading generated proto definitions, but it looks like a NULL struct dereference:

```
Segmentation fault at 0x00000000000090
```

```
-- Control frame information -----------------------------------------------
c:0138 p:---- s:0730 e:000729 CFUNC  :build
```

```
-- Machine register context ------------------------------------------------
 RIP: 0x00005566edca511f RBP: 0x00007f2530f2b188 RSP: 0x00007ffd8bd9fed0
 RAX: 0x0000000000000012 RBX: 0x00005566eebbbac0 RCX: 0x000000000000003d
 RDX: 0x000000000000013d RDI: 0x00005566eebbbac0 RSI: 0x0000000000000000
  R8: 0x00005566f1ef6ad0  R9: 0x0000000000000001 R10: 0x00005566f312d970
 R11: 0x00005566eebbbac0 R12: 0x00007f2530f28000 R13: 0x0000000000000004
 R14: 0x00007ffd8bda0010 R15: 0xcccccccccccccccd EFL: 0x0000000000010202
```

**Comment by pherl** [comment]
Can you verify if this has been fixed in 2.4.1? There was a bug related to maps implementation in multi-thread.

**Comment by tarcieri** [comment]
I assume you mean Ruby/MRI 2.4.1? I do happen to be using Ruby 2.4.0...

**Comment by TeBoring** [comment]
3.4.1?

**Comment by tarcieri** [comment]
Aah, this was on 3.4.1 which got yanked today. I upgraded to 3.4.1.1. Guess I'll see if it still has problems

**Comment by TeBoring** [comment]
How can it be reproduced?

**Comment by tarcieri** [comment]
I cannot reproduce it myself. It only occurred once.

**Comment by apolcyn** [comment]
I think I am seeing the same thing.

I'm seeing this reproduce 100% of the time with the following repro:

starting with this proto file:

```
syntax = "proto3";
message Empty {}
```

and generating with:

```
$PROTOC -I . empty.proto --ruby_out=.
```

that generates this ruby-protobuf code:

```
# Generated by the protocol buffer compiler.  DO NOT EDIT!
# source: empty.proto

require 'google/protobuf'

Google::Protobuf::DescriptorPool.generated_pool.build do
  add_message "Empty" do
  end
end

Empty = Google::Protobuf::DescriptorPool.generated_pool.lookup("Empty").msgclass
```


The the repro ruby script looks like:

```
GC.stress = 0x01 | 0x04
require_relative './empty_pb'
```

this is using `google-protobuf (3.4.1.1 x86_64-linux)` with `ruby 2.4.0 on x86_64-linux`

**Comment by TeBoring** [comment]
Thanks for the reproduce! I will take a look.

**Comment by TeBoring** [comment]
Fixed in #3708.
